File: java/p2/src/p2/types/operator/IndexJoin.java
package p2.types.operator;

import java.util.ArrayList;
import java.util.List;

import p2.lang.plan.Predicate;
import p2.lang.plan.Variable;
import p2.types.basic.Schema;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.exception.BadKeyException;
import p2.types.exception.P2RuntimeException;
import p2.types.table.Index;
import p2.types.table.Key;

public class IndexJoin extends Join {
	
	private Key lookupKey;
	
	private Index index;
	
	public IndexJoin(Predicate predicate, Schema input, Key lookupKey, Index index) {
		super(predicate, input);
		this.lookupKey = lookupKey;
		this.index = index;
	}
	
	@Override
	public String toString() {
		return "INDEX JOIN PREDICATAE[" + predicate.toString() + "]";
	}
	
	@Override
	public TupleSet evaluate(TupleSet tuples) throws P2RuntimeException {
		TupleSet result = new TupleSet();
		for (Tuple outer : tuples) {
			for (Tuple inner : this.index.lookup(this.lookupKey, outer)) {
				if (validate(outer, inner)) {
					inner.schema(this.predicate.schema().clone());
					Tuple join = outer.join(inner);
					if (join != null) {
						result.add(join);
					}
				}
			}
		}
		return result;
	}
}

File: lincoln/ruby/lib/types/operator/index_join.rb
require 'lib/types/operator/join'
class IndexJoin < Join
	
	def initialize (context, predicate, input, lookupKey, index)
		super(context, predicate, input)
		@lookupKey = lookupKey
		@index = index
	end
	
	def to_s
		return "INDEX JOIN: PREDICATE[" + @predicate.to_s + "]"
	end
	
	def evaluate(tuples)
		result = TupleSet.new("ij_out", nil)
		#puts tuples.name
    
		tuples.each do |outer|
  		# require 'ruby-debug'; debugger if outer.schema.name == 'clock'
			if tuples.name.to_s.eql?("global::MyPrimaryExpression") then
				## require 'ruby-debug'; debugger
			end
			#print "TEST #{outer}\n"
			@index.lookup_kt(@lookupKey, outer).each do |inner| 
				#print "TEST #{inner}\n"
				if validate(outer, inner) then
					#puts "found a tup for " + @index.table.name.name.to_s
## require 'ruby-debug'; debugger
					inner.schema = @predicate.schema.clone
					jointup = outer.join(inner)
					result << jointup unless jointup.nil?
				else 
					#nprint "NO tup for #{@index.table.name.name.to_s}\n"
					## require 'ruby-debug'; debugger
				end
			end
		end
                if result.size == 0 then
		   ## require 'ruby-debug';debugger
                   #puts "no tups for " + @index.table.name.name
                end
		return result
	end
end
